# For the first part, I am providing two separate solutions:

## 1) the first solution is to scape the saved html webpage as a file using my own method
## 2) the second solution is to use Pandas

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

import geocoder

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


This first part uses a generator (generating function) to return the lines in the html file based on the start line and stop line

The second function strips off the html text

In [2]:
def extract_data_lines(filename,start_text,end_text):
    """
    open 'filename', and yield the lines between
    the line that contains, 'start_text' and the
    line that contains 'end_text'
    """
    started = False
    ended = False
    with open(filename,encoding="utf8") as fh: #this means that after it is done reading the file it will close it automatically.
        for line in fh:
            #print(line)
            if start_text in line:
                started=True
                #print(line)
                yield line
            elif (started):
                if end_text in line:
                    ended=True
                    yield line
                    break #the break gets out of the loop over lines in the file
                elif (not ended):
                    yield line
                    
def rm_html_jnk(var):
    if ('title=' in var):
        ii = var.find('">')
        ij = var.find('</a>')
        var = var[ii+2:ij]
        return var
    else:
        if ('<td>' in var) and ('</td>' in var):
            ii = var.find('<td>')
            ij = var.find('</td>')
            var = var[ii+4:ij]
            return var
        if ('<td>' in var) and ('</td>' not in var):
            ii = var.find('<td>')
            var = var[ii+4:]
            return var.split('\n')[0]

In [3]:
filename = 'List_of_postal_codes_of_Canada_M.html'
start_text = '<td>M1A</td>'
end_text = '</tbody><tfoot></tfoot></table>'

data = {'Postcode': [],'Borough': [], 'Neighborhood': []}
lst = ['Postcode','Borough','Neighborhood']
post = {}
tr = False

counter = 0 #since the html file splits the row into three separate parts for each column

idx = 0     #since there are repeat postalcodes this index variable keeps track of unique one in a dictionary

temp = []   #this list gets reset every time it hits a new row of variables

same = False #variable to say whether or not new postalcode is already in the dictionary

for line in extract_data_lines(filename,start_text,end_text):
    if end_text not in line:
        #print(temp)
        if '<tr>' in line or '</tr>' in line: #these tags mark the beginning/end of the column variables for the row
            counter = 0
            temp = []
            same = False
        else:
            trim_line = rm_html_jnk(line)
            #temp.append(trim_line)
            if(counter == 0):
                temp.append(trim_line)
                if trim_line in post:
                    same = True
                else:
                    post[trim_line]=idx
                    data[lst[counter]].append(trim_line)
                    idx+=1
            elif(counter == 1):
                temp.append(trim_line)
                if not same:
                    data[lst[counter]].append(rm_html_jnk(line))
            elif(counter == 2):
                temp.append(trim_line) 
                #Below I reassign the Neighbourhood variable if the Borough variable is set
                if trim_line =='Not assigned' and temp[1] != 'Not assigned':
                    trim_line = temp[1]
                    print('resetting neighborhood from {} to {}'.format(temp[2],trim_line))
                if same: 
                    #We can just add the neighborhoods to all the others of the same postalcode
                    data[lst[counter]][post[temp[0]]] = data[lst[counter]][post[temp[0]]]+', '+trim_line
                else:
                    data[lst[counter]].append(trim_line)
            counter+=1

df = pd.DataFrame(data)

resetting neighborhood from Not assigned to Queen's Park


### As you can see below, this was properly read in except for the fact that we need to drop the rows with 'Not assigned' in the 'Borough' column

In [4]:
print(df.shape)
df.head()

(180, 3)


Postcode           Borough               Neighborhood
0      M1A      Not assigned               Not assigned
1      M2A      Not assigned               Not assigned
2      M3A        North York                  Parkwoods
3      M4A        North York           Victoria Village
4      M5A  Downtown Toronto  Harbourfront, Regent Park

In [5]:
df_dropped = df[df['Borough'] != 'Not assigned']

In [6]:
df_dropped.reset_index(drop=True,inplace=True)

In [7]:
df_dropped.head()

Postcode           Borough                      Neighborhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
3      M6A        North York  Lawrence Heights, Lawrence Manor
4      M7A      Queen's Park                      Queen's Park

In [8]:
df_dropped.shape

(103, 3)

## Now for the method to use Pandas to read in the table from the url using pd.read_url()

In [9]:
weburl = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
new_df = pd.read_html(weburl)[0]  #The [0] at the end is necessary to pull the first part of the data read in which is the table

In [10]:
print(new_df.shape)
new_df.head()

(288, 3)


Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [11]:
new_df_drop = new_df[new_df['Borough'] != 'Not assigned']
new_df_drop.reset_index(drop=True,inplace=True)

In [12]:
def set_same_borough(x):
    for i in range(len(x['Neighbourhood'])):
        if x['Neighbourhood'][i]=='Not assigned':
            x['Neighbourhood'][i]=x['Borough'][i]
    return x

#I am going to set the same neighborhood as the borough if the neighborhood is not assigned
new_df_drop = set_same_borough(new_df_drop)
print(new_df_drop[new_df_drop['Neighbourhood']=='Not assigned'])

Empty DataFrame
Columns: [Postcode, Borough, Neighbourhood]
Index: []


C:\Users\KevinDWilk\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [13]:
new_df_drop.shape

(211, 3)

In [14]:
new_df_group = new_df_drop.groupby('Postcode')

In [15]:
def join_nghds(x):
    x['Neighbourhood'] = ', '.join(x['Neighbourhood'])
    return x
final_df=new_df_group.apply(join_nghds)

In [16]:
final_df.head()

Postcode           Borough                     Neighbourhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
3      M5A  Downtown Toronto         Harbourfront, Regent Park
4      M6A        North York  Lawrence Heights, Lawrence Manor

In [17]:
final_df.drop_duplicates(subset='Postcode',keep='first',inplace=True)
final_df.head()

Postcode           Borough                     Neighbourhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
4      M6A        North York  Lawrence Heights, Lawrence Manor
6      M7A      Queen's Park                      Queen's Park

In [18]:
print('Note it is the same as the previous method...YAY!')
final_df.shape

Note it is the same as the previous method...YAY!


(103, 3)

# Now on to the second part

## Remember the goal of this part is to read in the coordinates information and make the table.

In [20]:
lat_coords = []
long_coords = []
for postal_code in final_df['Postcode']:
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
    print('Finished for postal code: {}'.format(postal_code))
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    lat_coords.append(latitude)
    long_coords.append(longitude)

final_df['Latitude']=lat_coords
final_df['Longitude']=long_coords

KeyboardInterrupt: 

Since this method doesn't seem to be working, I am going to just load in the datasets and merge them.

In [27]:
coords_df = pd.read_csv('Geospatial_Coordinates.csv')

In [28]:
coords_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [26]:
coords_df.shape

(103, 3)

They have the same length (103 rows), so I will assume that it has all the same Postal Codes that I have

In [31]:
coords_df.rename(columns={'Postal Code':'Postcode'},inplace=True)

In [33]:
joined_df = pd.merge(final_df,coords_df, on='Postcode',how='inner')

In [37]:
joined_df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
30       M5H  Downtown Toronto   
31       M6H      West Toronto   
32       M1J       Scarborough   
33       M2J        North York   
34       M3J        North York   
35       M4J         East York   
36       M5J  Downtown Toronto   
37       M6J      West Toronto   
38       M1K       Scarborough   
39       M2K        North York   
40       M3K        North York   
41       M4K      East Toronto   
42       M5K  Downtown Toronto   
43       M6K      West Toronto   
44       M1L       Scarborough   
45       M2L        North York   
46       M3L        North York   
47       M4L      East Toronto   
48       M5L  Downtown Toronto   
49       M6L        North York   
50       M9L        North York   
51       M1M       Scarborough   
52       M2M        North York   
53       M3M        North York   
54       M4M      East Toronto   
55       M5M        North York   
56       M6M              York   
57       M9M        North York   
58       M1N       Scarborough   
59       M2N        North York   
60       M3N        North York   
61       M4N   Central Toronto   
62       M5N   Central Toronto   
63       M6N              York   
64       M9N              York   
65       M1P       Scarborough   
66       M2P        North York   
67       M4P   Central Toronto   
68       M5P   Central Toronto   
69       M6P      West Toronto   
70       M9P         Etobicoke   
71       M1R       Scarborough   
72       M2R        North York   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Harbourfront, Regent Park  43.654260 -79.360636  
3                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.38

We've now made the table for the coordinates for each postal code...

# Now for the third part.
## In this third part we apply clustering algorithms to the neighborhoods of Toronto, Canada

In [38]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Ontario are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, Ontario are 43.653963, -79.387207.


In [39]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(joined_df['Latitude'], joined_df['Longitude'], 
                                           joined_df['Borough'], joined_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [40]:
CLIENT_ID = 'UZNFARAB5RPK3VRKCEPIBJIU0YUBE4EMDMUOWL3FK2OBNLK4' # your Foursquare ID
CLIENT_SECRET = 'R3KLBEUMMU5A3FTOX4WW1SRJHKED1SQXGMSYTFXHFJ1PEXRE' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: UZNFARAB5RPK3VRKCEPIBJIU0YUBE4EMDMUOWL3FK2OBNLK4
CLIENT_SECRET:R3KLBEUMMU5A3FTOX4WW1SRJHKED1SQXGMSYTFXHFJ1PEXRE


### Since there are a lot of boroughs, I will restrict this to just boroughs with Toronto in the name.

In [41]:
toronto_data = joined_df[joined_df['Borough'].str.contains('Toronto')]

In [42]:
toronto_data.head()

Postcode           Borough              Neighbourhood   Latitude  Longitude
2       M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260 -79.360636
9       M5B  Downtown Toronto   Ryerson, Garden District  43.657162 -79.378937
15      M5C  Downtown Toronto             St. James Town  43.651494 -79.375418
19      M4E      East Toronto                The Beaches  43.676357 -79.293031
20      M5E  Downtown Toronto                Berczy Park  43.644771 -79.373306

In [43]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], 
                                           toronto_data['Borough'], toronto_data['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [50]:
toronto_data[toronto_data['Borough']=='Downtown Toronto']

Postcode           Borough  \
2       M5A  Downtown Toronto   
9       M5B  Downtown Toronto   
15      M5C  Downtown Toronto   
20      M5E  Downtown Toronto   
24      M5G  Downtown Toronto   
25      M6G  Downtown Toronto   
30      M5H  Downtown Toronto   
36      M5J  Downtown Toronto   
42      M5K  Downtown Toronto   
48      M5L  Downtown Toronto   
80      M5S  Downtown Toronto   
84      M5T  Downtown Toronto   
87      M5V  Downtown Toronto   
91      M4W  Downtown Toronto   
92      M5W  Downtown Toronto   
96      M4X  Downtown Toronto   
97      M5X  Downtown Toronto   
99      M4Y  Downtown Toronto   

                                        Neighbourhood   Latitude  Longitude  
2                           Harbourfront, Regent Park  43.654260 -79.360636  
9                            Ryerson, Garden District  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
20                                        Berczy Park  43.644771 -79.373306  
24                                 Central Bay Street  43.657952 -79.387383  
25                                           Christie  43.669542 -79.422564  
30                           Adelaide, King, Richmond  43.650571 -79.384568  
36  Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752  
42           Design Exchange, Toronto Dominion Centre  43.647177 -79.381576  
48                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
80                     Harbord, University of Toronto  43.662696 -79.400049  
84          Chinatown, Grange Park, Kensington Market  43.653206 -79.400049  
87  CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420  
91                                           Rosedale  43.679563 -79.377529  
92                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
96                        Cabbagetown, St. James Town  43.667967 -79.367675  
97             First Canadian Place, Underground city  43.648429 -79.382280  
99                               Church and Wellesley  43.665860 -79.383160

In [55]:
toronto_data[toronto_data['Borough']=='Downtown Toronto'].reset_index(drop=True).loc[0,'Neighbourhood'].split(',')[0]

'Harbourfront'

In [54]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

2                             Harbourfront, Regent Park
9                              Ryerson, Garden District
15                                       St. James Town
20                                          Berczy Park
24                                   Central Bay Street
25                                             Christie
30                             Adelaide, King, Richmond
36    Harbourfront East, Toronto Islands, Union Station
42             Design Exchange, Toronto Dominion Centre
48                       Commerce Court, Victoria Hotel
80                       Harbord, University of Toronto
84            Chinatown, Grange Park, Kensington Market
87    CN Tower, Bathurst Quay, Island airport, Harbo...
91                                             Rosedale
92                      Stn A PO Boxes 25 The Esplanade
96                          Cabbagetown, St. James Town
97               First Canadian Place, Underground city
99                                 Church and We

In [57]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### After using the function above for the call to the API, we can now call it using our Toronto data frame

In [58]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide, King, Richmond
Dovercourt Village, Dufferin
Harbourfront East, Toronto Islands, Union Station
Little Portugal, Trinity
The Danforth West, Riverdale
Design Exchange, Toronto Dominion Centre
Brockton, Exhibition Place, Parkdale Village
The Beaches West, India Bazaar
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North, Forest Hill West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
Harbord, University of Toronto
Runnymede, Swansea
Moore Park, Summerhill East
Chinatown, Grange Park, Kensington Market
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown, St. James Town
Fir

In [61]:
print(toronto_venues.shape)
toronto_venues.head()

(1717, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront, Regent Park               43.65426              -79.360636   
1  Harbourfront, Regent Park               43.65426              -79.360636   
2  Harbourfront, Regent Park               43.65426              -79.360636   
3  Harbourfront, Regent Park               43.65426              -79.360636   
4  Harbourfront, Regent Park               43.65426              -79.360636   

                                             Venue  Venue Latitude  \
0                                 Roselle Desserts       43.653447   
1                                    Tandem Coffee       43.653559   
2  Toronto Cooper Koo Family Cherry St YMCA Centre       43.653191   
3                              Body Blitz Spa East       43.654735   
4                               Morning Glory Cafe       43.653947   

   Venue Longitude        Venue Category  
0       -79.362017                Bakery  
1       -79.361809           Coffee Shop  
2       -79.357947  Gym / Fitness Center  
3       -79.359874                   Spa  
4       -79.361149        Breakfast Spot

In [62]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            57   
Brockton, Exhibition Place, Parkdale Village                           24   
Business Reply Mail Processing Centre 969 Eastern                      15   
CN Tower, Bathurst Quay, Island airport, Harbou...                     17   
Cabbagetown, St. James Town                                            45   
Central Bay Street                                                     89   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   88   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        9   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     16   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         35   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              52   
High Park, The Junction South                                          24   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               66   
Moore Park, Summerhill East                                             3   
North Toronto West                                                     20   
Parkdale, Roncesvalles                                                 15   
Rosedale                                                                5   
Roselawn                                                                2   
Runnymede, Swansea                                                     35   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        98   
Studio District                                                        39   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             6   
The Beaches West, India Bazaar                                         20   
The Danforth West, Riverdale                                           43   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             57   
Brockton, Exhibition Place, Parkdale Village                            24   
Business Reply Mail Processing Centre 969 Eastern                       15   
CN Tower, Bathurst Quay, Island airport, Harbou...                      17   
Cabbagetown, St. James Town                                             45   
Central Bay Street                                                      89   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    88   
Commerce Court, Victoria Hotel     

In [63]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 242 uniques categories.


In [64]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               1        0                0         0             0   

   Burrito Place  Bus Line  Bus Stop  Business Service  Butcher  Café  \
0              0         0         0                 0        0     0   
1              0         0         0                 0        0     0   
2              0         0         0                 0        0     0   
3              0         0     

In [65]:
toronto_onehot.shape

(1717, 242)

In [66]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.066667   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.011236   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.022727   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.000000   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.019231   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.015152   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.050000   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.025641   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023256   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.058824            0.058824      0.058824   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.011364  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.000000  0.000000            0.000000      0.000000   
17           0.000000  0.000000            0.000000      0.000000   

In [67]:
toronto_grouped.shape

(38, 242)

In [68]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2            Beer Bar  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Exhibition Place, Parkdale Village----
                   venue  freq
0                   Café  0.12
1         Breakfast Spot  0.08
2  Performing Arts Venue  0.08
3            Coffee Shop  0.08
4   Caribbean Restaurant  0.04


----Business Reply Mail Processing Centre 969 Eastern----
           venue  freq
0    Yoga Studio  0.07
1  Auto Workshop  0.07
2     Comic Shop  0.07
3     Restaurant  0.07
4     Skate Park  0.07


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge

In [69]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [70]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                  Café            Steakhouse                    Bar   
1          Cocktail Bar    Seafood Restaurant            Cheese Shop   
2           Coffee Shop        Breakfast Spot  Performing Arts Venue   
3        Farmers Market            Restaurant             Smoke Shop   
4        Airport Lounge      Airport Terminal                  Plane   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant   American Restaurant                 Hotel   
1            Steakhouse              Beer Bar        Farmers Market   
2  Caribbean Restaurant               Stadium         Burrito Place   
3         Auto Workshop               Brewery  Fast Food Restaurant   
4                   Bar           Coffee Shop      Sculpture Garden   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant      Sushi Restaurant       Asian Restaurant  
1                Bakery                  Café              Jazz Club  
2        Sandwich Place            Restaurant           Climbing Gym  
3                Garden         Garden Center     Light Rail Station  
4              Boutique         Boat or Ferry        Harbor / Marina

## Cluster Neighborhoods

In [80]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=1).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:20] 

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 3])

In [81]:
toronto_grouped_clustering.head()

Yoga Studio  Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0     0.000000                0.0  0.000000            0.000000      0.000000   
1     0.000000                0.0  0.000000            0.000000      0.000000   
2     0.000000                0.0  0.000000            0.000000      0.000000   
3     0.066667                0.0  0.000000            0.000000      0.000000   
4     0.000000                0.0  0.058824            0.058824      0.058824   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.000000         0.000000          0.000000                 0.03   
1        0.000000         0.000000          0.000000                 0.00   
2        0.000000         0.000000          0.000000                 0.00   
3        0.000000         0.000000          0.000000                 0.00   
4        0.117647         0.176471          0.117647                 0.00   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0     0.010000        0.01                  0.0   
1           0.0       0.0     0.017544        0.00                  0.0   
2           0.0       0.0     0.000000        0.00                  0.0   
3           0.0       0.0     0.000000        0.00                  0.0   
4           0.0       0.0     0.000000        0.00                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0              0.03                 0.0       0.000000   0.000000         0.0   
1              0.00                 0.0       0.000000   0.017544         0.0   
2              0.00                 0.0       0.000000   0.000000         0.0   
3              0.00                 0.0       0.066667   0.000000         0.0   
4              0.00                 0.0       0.000000   0.000000         0.0   

   Bagel Shop    Bakery  Bank       Bar  Baseball Stadium  Basketball Stadium  \
0    0.000000  0.020000   0.0  0.040000               0.0            0.000000   
1    0.017544  0.035088   0.0  0.000000               0.0            0.017544   
2    0.000000  0.041667   0.0  0.041667               0.0            0.000000   
3    0.000000  0.000000   0.0  0.000000               0.0            0.000000   
4    0.000000  0.000000   0.0  0.058824               0.0            0.000000   

      Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0  0.000000              0.0  0.000000         0.0                 0.0   
1  0.017544              0.0  0.035088         0.0                 0.0   
2  0.000000              0.0  0.000000         0.0                 0.0   
3  0.000000              0.0  0.000000         0.0                 0.0   
4  0.000000              0.0  0.000000         0.0                 0.0   

     Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0  0.000000         0.0       0.000000       0.01  0.000000   
1  0.017544         0.0       0.000000       0.00  0.000000   
2  0.000000         0.0       0.000000       0.00  0.000000   
3  0.000000         0.0       0.000000       0.00  0.000000   
4  0.000000         0.0       0.058824       0.00  0.058824   

   Brazilian Restaurant  Breakfast Spot   Brewery  Bubble Tea Shop  Building  \
0                  0.01        0.020000  0.000000              0.0      0.01   
1                  0.00        0.017544  0.000000              0.0      0.00   
2                  0.00        0.083333  0.000000              0.0      0.00   
3                  0.00        0.000000  0.066667              0.0      0.00   
4                  0.00        0.000000  0.000000              0.0      0.00   

   Burger Joint  Burrito Place  Bus Line  Bus Stop  Business Service  Butcher  \
0          0.03       0.010000       0.0       0.0               0.0      0.0   
1          0.00       0.000000       0.0       0.0               0.0      0.0   
2          0.00       0.041667       0.0       0.0               0.0      0.0   

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [85]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

ValueError: cannot insert Cluster Labels, already exists